### PR TITLE
add dnnl patterns for byoc, along with their test units

### PR DESCRIPTION
**Pull request by crazydemo**
The test unit structure is inspired by `test_tensorrt.py`.
I consider this PR will be the next PR that is to be submitted to `apahce/tvm/master`.

### CODE REVIEW

**Comment by yangulei** [comment]
@Zhennan

**Comment by yangulei** [comment]
@ZhennanQin

File: python/tvm/relay/op/contrib/dnnl.py [hunk 1]
```diff
@@ -76,12 +78,79 @@ def make_pattern(with_bias=True):
...
-    conv2d_bias_relu_pat = ("dnnl.conv2d_bias_relu", make_pattern(with_bias=True))
-    conv2d_relu_pat = ("dnnl.conv2d_relu", make_pattern(with_bias=False))
-    dnnl_patterns = [conv2d_bias_relu_pat, conv2d_relu_pat]
+    conv2d_bias_relu_pat = (
+        "dnnl.conv2d_bias_relu",
+        make_conv_pattern(with_bias=True, with_relu=True),
+    )
+    conv2d_bias_pat = ("dnnl.conv2d_bias", make_conv_pattern(with_bias=True, with_relu=False))
+    conv2d_relu_pat = ("dnnl.conv2d_relu", make_conv_pattern(with_bias=False, with_relu=True))
+    dense_bias_pat = ("dnnl.dense_bias", make_dense_pattern(with_bias=True, with_relu=False))
```
Comment (ciyongch) [inline python/tvm/relay/op/contrib/dnnl.py]: Is it easy to extend the pattern list for various elementwise post-ops other than relu (for example, tanh/sigmoid/mul, etc.)?

File: python/tvm/relay/op/contrib/dnnl.py [hunk 2]
```diff
@@ -76,12 +78,79 @@ def make_pattern(with_bias=True):
...
+        conv2d_bias_relu_pat,
+        conv2d_bias_pat,
+        conv2d_relu_pat,
+        dense_bias_pat,
+    ]
     return dnnl_patterns
+
+
+def partition_for_dnnl(
+    mod,
```
Comment (ciyongch) [inline python/tvm/relay/op/contrib/dnnl.py]: format? Keep the two args in the same line of function declaration.
I'm wondering if there's any type check for python code in TVM?

File: python/tvm/relay/op/contrib/dnnl.py [hunk 3]
```diff
@@ -76,12 +78,79 @@ def make_pattern(with_bias=True):
...
+            # fold consecutive add ops to simplify pattern `conv2d-bias_add-bn-relu`
+            transform.SimplifyExpr(),
+            transform.FoldConstant(),
+            transform.MergeComposite(pattern_table()),
+            transform.AnnotateTarget("dnnl"),
+            transform.MergeCompilerRegions(),
+            transform.PartitionGraph(),
+        ]
+    )
+    with tvm.transform.PassContext(opt_level=3):
```
Comment (ciyongch) [inline python/tvm/relay/op/contrib/dnnl.py:215]: what's the difference result when the `opt_level` equals to values like 0/1/2? do you think user need this option when creating partition for dnnl backend?

Comment (crazydemo) [reply python/tvm/relay/op/contrib/dnnl.py:215]: if `opt_level=1 / 2`, some passes like alter_op_layout will be disable, then it cannot pass the test cases, `opt_level=3` is the default setting.

File: python/tvm/relay/op/contrib/dnnl.py (hunk 2 above)
Comment (crazydemo) [reply python/tvm/relay/op/contrib/dnnl.py]: fixed. TVM has C++ and python lint, all lint has been passes.

File: python/tvm/relay/op/contrib/dnnl.py (hunk 1 above)
Comment (crazydemo) [reply python/tvm/relay/op/contrib/dnnl.py]: It is easy to extend on dnnl codegen, but can be a bit hard on tvm, as it need to registster all ops first. 
I will try to figure out.

Comment (crazydemo) [reply python/tvm/relay/op/contrib/dnnl.py]: I have updated the related code. 
If we want to extend the pattern list, we need to extend the `elt_list`.
And we also need to add pattern in `dnnl_json_runtime.cc`

File: python/tvm/relay/op/contrib/dnnl.py [hunk 4]
```diff
@@ -76,12 +81,83 @@ def make_pattern(with_bias=True):
...
+def make_dnnl_pattern(op, with_bias, with_eltwise):
+    pat_name = "dnnl." + op
+    pat_name += "_bias" if with_bias else ""
+    pat_name += ("_" + with_eltwise.split(".")[-1]) if with_eltwise else ""
+    if op == "conv2d":
+        return (pat_name, make_conv_pattern(with_bias, with_eltwise))
+    elif op == "dense":
+        return (pat_name, make_dense_pattern(with_bias, with_eltwise))
+    else:
+        return None
```
Comment (ciyongch) [inline python/tvm/relay/op/contrib/dnnl.py]: suggest to add some warning msg like `Currently, only conv2d and dense op are supported, but got ()".format(op).`

File: python/tvm/relay/op/contrib/dnnl.py [hunk 5]
```diff
@@ -76,12 +81,83 @@ def make_pattern(with_bias=True):
...
 def pattern_table():
-    conv2d_bias_relu_pat = ("dnnl.conv2d_bias_relu", make_pattern(with_bias=True))
-    conv2d_relu_pat = ("dnnl.conv2d_relu", make_pattern(with_bias=False))
-    dnnl_patterns = [conv2d_bias_relu_pat, conv2d_relu_pat]
+    elt_list = ["nn.relu", "tanh", "sigmoid", None]
+    dnnl_patterns = []
+    for with_bias in [True, False]:
+        for elt in elt_list:
+            if not with_bias and not elt:
+                return dnnl_patterns
```
Comment (ciyongch) [inline python/tvm/relay/op/contrib/dnnl.py:176]: no need these without "None"  in elt_list.

**Review by ciyongch — COMMENTED** [review]
Overall looks good:)

File: python/tvm/relay/op/contrib/dnnl.py (hunk 5 above)
Comment (crazydemo) [reply python/tvm/relay/op/contrib/dnnl.py:176]: I need this to construct patterns like `conv-bias`, who has no activation attached.

File: python/tvm/relay/op/contrib/dnnl.py (hunk 4 above)
Comment (crazydemo) [reply python/tvm/relay/op/contrib/dnnl.py]: warning log has been added.

File: python/tvm/relay/op/contrib/dnnl.py (hunk 5 above)
Comment (ciyongch) [reply python/tvm/relay/op/contrib/dnnl.py:176]: Ok, then it should be fine.

**Review by ciyongch — APPROVED** [review]
LGTM.

**Pull request closed without merging**